src/runtime-rs/crates/hypervisor/src/dragonball/vmm_instance.rs

```diff
@@ -0,0 +1,334 @@
+        InstanceInfo, InstanceState, VirtioNetDeviceConfigInfo, VmmAction, VmmActionError, VmmData,
+        VmmRequest, VmmResponse, VmmService, VsockDeviceConfigInfo,
+    },
+    vm::VmConfigInfo,
+    Vmm,
+};
+use nix::sched::{setns, CloneFlags};
+use seccompiler::BpfProgram;
+use vmm_sys_util::eventfd::EventFd;
+
+use crate::ShareFsOperation;
+
+pub enum Request {
+    Sync(VmmAction),
+}
+
+const DRAGONBALL_VERSION: &str = env!("CARGO_PKG_VERSION");
+const REQUEST_RETRY: u32 = 500;
+
+pub struct VmmInstance {
+    /// VMM instance info directly accessible from runtime
+    vmm_shared_info: Arc<RwLock<InstanceInfo>>,
+    to_vmm: Option<Sender<VmmRequest>>,
+    from_vmm: Option<Receiver<VmmResponse>>,
+    to_vmm_fd: EventFd,
+    seccomp: BpfProgram,
+    vmm_thread: Option<thread::JoinHandle<Result<i32>>>,
+}
+
+impl VmmInstance {
+    pub fn new(id: &str) -> Self {
+        let vmm_shared_info = Arc::new(RwLock::new(InstanceInfo::new(
+            String::from(id),
+            DRAGONBALL_VERSION.to_string(),
+        )));
+
+        let to_vmm_fd = EventFd::new(libc::EFD_NONBLOCK)
+            .unwrap_or_else(|_| panic!("Failed to create eventfd for vmm {}", id));
+
+        VmmInstance {
+            vmm_shared_info,
+            to_vmm: None,
+            from_vmm: None,
+            to_vmm_fd,
+            seccomp: vec![],
+            vmm_thread: None,
+        }
+    }
+
+    pub fn get_shared_info(&self) -> Arc<RwLock<InstanceInfo>> {
+        self.vmm_shared_info.clone()
+    }
+
+    fn set_instance_id(&mut self, id: &str) {
+        let share_info_lock = self.vmm_shared_info.clone();
+        share_info_lock.write().unwrap().id = String::from(id);
+    }
+
+    pub fn get_vcpu_tids(&self) -> Vec<(u8, u32)> {
+        let info = self.vmm_shared_info.clone();
+        let result = info.read().unwrap().tids.clone();
+        result
+    }
+
+    pub fn run_vmm_server(&mut self, id: &str, netns: Option<String>) -> Result<()> {
+        let kvm = OpenOptions::new().read(true).write(true).open("/dev/kvm")?;
+
+        let (to_vmm, from_runtime) = channel();
+        let (to_runtime, from_vmm) = channel();
+
+        self.set_instance_id(id);
+
+        let vmm_service = VmmService::new(from_runtime, to_runtime);
+
+        self.to_vmm = Some(to_vmm);
+        self.from_vmm = Some(from_vmm);
+
+        let api_event_fd2 = self.to_vmm_fd.try_clone().expect("Failed to dup eventfd");
+        let vmm = Vmm::new(
+            self.vmm_shared_info.clone(),
+            api_event_fd2,
+            self.seccomp.clone(),
+            self.seccomp.clone(),
+            Some(kvm.into_raw_fd()),
+        )
+        .expect("Failed to start vmm");
+
+        self.vmm_thread = Some(
+            thread::Builder::new()
+                .name("vmm_master".to_owned())
+                .spawn(move || {
+                    || -> Result<i32> {
+                        debug!(sl!(), "====> run vmm thread start");
```

Comment:
I personally find "eye catchers" like `====>` visually distracting and I don't think they are necessary.

Reply:
Thanks for your review, already removed.